lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,712 @@
+        local test_horizontal = not (row > data.row + data.row_span - 1 or data.row > row + row_span - 1)
+        -- If one rectangular widget is above other
+        local test_vertical   = not (col > data.col + data.col_span - 1 or data.col > col + col_span - 1)
+        if test_horizontal and test_vertical then
+            table.insert(ret, index)
+        end
+    end
+    table.sort(ret)
+    return #ret > 0 and ret or nil
+end
+
+--- Get (and update) the number of rows and columns occupied by the widgets in the grid
+-- @treturn number,number the number of rows and columns
+function grid:get_shape()
+    local num_rows, num_cols = 0, 0
+    if self._private.forced_num_rows then
+        num_rows = self._private.forced_num_rows
+    end
+    if self._private.forced_num_cols then
+        num_cols = self._private.forced_num_cols
+    end
+
+    for _, data in ipairs(self._private.widgets) do
+        num_rows = math.max(num_rows, data.row + data.row_span - 1)
+        num_cols = math.max(num_cols, data.col + data.col_span - 1)
+    end
+    self._private.num_rows = num_rows
+    self._private.num_cols = num_cols
+    return num_rows, num_cols
+end
+
+
+--- Add a widget to the grid layout at specific coordinate
+--@DOC_wibox_layout_grid_add_EXAMPLE@
+-- @param child Widget that should be added
+-- @param number row Row number for the top left corner of the widget
+-- @param number col Column number for the top left corner of the widget
+-- @tparam number row_span The number of rows the widget spans (default to 1)
+-- @tparam number col_span The number of columns the widget spans (default to 1)
+-- @treturn boolean index If the operation is successful
+function grid:add_widget_at(child, row, col, row_span, col_span)
+    local rspan = (row_span and row_span > 0) and row_span or 1
+    local cspan = (col_span and col_span > 0) and col_span or 1
+
+    -- check if the object is a widget
+    base.check_widget(child)
+
+    -- test if the new widget superpose with existing ones
+    local superpose = find_widgets_by_coordinates(self._private.widgets, row, col, rspan, cspan)
+    if superpose then
+        -- TODO: add message, remove other widgets?
+        return false
+    end
+
+    -- Add grid information attached to the widget
+    local child_data = {
+        index = #self._private.widgets + 1,  -- debug only
+        widget = child,
+        row = row,
+        col = col,
+        row_span = rspan,
+        col_span = cspan
+    }
+    table.insert(self._private.widgets, child_data)
+    matrix_add(self._private.matrix, child_data, row, col, rspan, cspan)
+
+    -- Update the row and column numbers
+    if self._private.num_rows < row + rspan - 1 then
+        self._private.num_rows = row + rspan - 1
+    end
+    if self._private.num_cols < col + cspan - 1 then
+        self._private.num_cols = col + cspan - 1
+    end
+
+    self:emit_signal("widget::layout_changed")
+    self:emit_signal("widget::redraw_needed")
```

Comment:
Why `widget::redraw_needed`? Removing a widget does not emit that signal and by symmetry both cases should emit the same signals.